### PR TITLE
Implement configuration logic 

### PR DESCRIPTION
**Pull request by Mitko-Kerezov**
Includes changes for
* Prop set/print commands
* Build
* LiveSync (both regular and companion)

TP Reference: [Support environment configurations](http://teampulse.telerik.com/view#item/308175)
Ping @rosen-vladimirov @TsvetanMilanov 

Tests are on the way!

### CODE REVIEW

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: appbuilder/project/project-base.ts [hunk 1]
```diff
@@ -40,5 +62,92 @@ export class Project implements Project.IProjectBase {
...
+					let allProjectFiles = this.$fs.enumerateFilesInDirectorySync(projectDir, (file: string, stat: IFsStats) => {
+						return ProjectBase.CONFIGURATION_FILE_SEARCH_PATTERN.test(file);
+					});
+
+					_.each(allProjectFiles, (configProjectFile: string) => {
+						let configMatch = path.basename(configProjectFile).match(ProjectBase.CONFIGURATION_FROM_FILE_NAME_REGEX);
+						if(configMatch && configMatch.length > 1) {
+							let configurationName = configMatch[1];
+							let configProjectContent = this.$fs.readJson(configProjectFile).wait();
+							this.configurationSpecificData[configurationName.toLowerCase()] = <any>_.merge(JSON.parse(JSON.stringify(this.projectData)), configProjectContent);
```
Comment (TsvetanMilanov) [inline appbuilder/project/project-base.ts]: If we want to support case sensitive project configurations we do not need this `.toLowerCase()`

File: appbuilder/project/project-base.ts [hunk 2]
```diff
@@ -1,29 +1,51 @@
+import {EOL} from "os";
 import Future = require("fibers/future");
-import { TARGET_FRAMEWORK_IDENTIFIERS } from "../../mobile/constants";
+import { TARGET_FRAMEWORK_IDENTIFIERS } from "../../constants";
 import * as path from "path";
-import { startPackageActivityNames } from "../../mobile/constants";
+import { startPackageActivityNames } from "../../constants";
+export abstract class ProjectBase implements Project.IProjectBase {
+	private static CONFIGURATION_FILE_SEARCH_PATTERN: RegExp = new RegExp(".*.abproject$", "i");
```
Comment (TsvetanMilanov) [inline appbuilder/project/project-base.ts]: If the user creates configuration file with name which contains invalid characters for JavaScript object property this regular expression will match it as valid configuration name. I think `^.(?:[A-z\d]+.)?abproject$` will be better.

File: declarations.d.ts [hunk 3]
```diff
@@ -491,10 +491,10 @@ interface ILiveSyncServiceBase {
 	 * @param {ILiveSyncData} data LiveSync data describing the LiveSync operation.
 	 * @param {string[]} filesToSync Files that have to be synced.
 	 * @param {Function} deviceFilesAction Custom action that has to be executed instead of just copying the files.
-	 * @param {any} liveSyncOptions Defines if the LiveSync operation is for Companion app.
+	 * @param {any} liveSyncOptions Additional options for LiveSyncing
```
Comment (TsvetanMilanov) [inline declarations.d.ts]: Change the param type from `any` to `ILiveSyncOptions`.

File: appbuilder/project/project-base.ts [hunk 4]
```diff
@@ -40,5 +62,92 @@ export class Project implements Project.IProjectBase {
...
+					if(this.$options.debug && !this.$fs.exists(debugProjectFile).wait()) {
+						this.$fs.writeJson(debugProjectFile, {}).wait();
+					}
+
+					let releaseProjectFile = path.join(projectDir, this.$projectConstants.RELEASE_PROJECT_FILE_NAME);
+					if(this.$options.release && !this.$fs.exists(releaseProjectFile).wait()) {
+						this.$fs.writeJson(releaseProjectFile, {}).wait();
+					}
+
+					let allProjectFiles = this.$fs.enumerateFilesInDirectorySync(projectDir, (file: string, stat: IFsStats) => {
```
Comment (TsvetanMilanov) [inline appbuilder/project/project-base.ts]: I think we don't need this variable, because we can enumerate all the files in the directory and directly check for project configuration matches using `ProjectBase.CONFIGURATION_FROM_FILE_NAME_REGEX`.

File: mobile/ios/device/ios-device-file-system.ts [hunk 5]
```diff
@@ -14,7 +14,7 @@ export class IOSDeviceFileSystem implements Mobile.IDeviceFileSystem {
 		private $mobileDevice: Mobile.IMobileDevice,
 		private $options: ICommonOptions) { }
 
-	public listFiles(devicePath: string): IFuture<void> {
+	public listFiles(devicePath: string, appIdentifier?: string): IFuture<any> {
```
Comment (TsvetanMilanov) [inline mobile/ios/device/ios-device-file-system.ts]: Do we need `appIdentifier` in this method?

File: mobile/android/android-device.ts [hunk 6]
```diff
@@ -100,6 +99,24 @@ export class AndroidDevice implements Mobile.IAndroidDevice {
 		return this.deviceInfo.type === "Emulator";
 	}
 
+	public getApplicationInfo(applicationIdentifier: string): IFuture<Mobile.IApplicationInfo> {
+		return ((): Mobile.IApplicationInfo => {
+			let files = this.fileSystem.listFiles("files/12590FAA-5EDD-4B12-856D-F52A0A1599F2", applicationIdentifier).wait(),
```
Comment (TsvetanMilanov) [inline mobile/android/android-device.ts]: Can you extract `12590FAA-5EDD-4B12-856D-F52A0A1599F2` in constant?

File: appbuilder/project/project-base.ts (hunk 2 above)
Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: `[.].+[.]abproject$`

Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: let's just remove this RegEx

File: appbuilder/project/project-base.ts [hunk 7]
```diff
@@ -40,5 +62,92 @@ export class Project implements Project.IProjectBase {
...
+	protected readProjectData(): IFuture<void> {
+		return (() => {
+			let projectDir = this.getProjectDir().wait();
+			this.setShouldSaveProject(false);
+			if(projectDir) {
+				let projectFilePath = path.join(projectDir, this.$projectConstants.PROJECT_FILE);
+				try {
+					let data = this.$fs.readJson(projectFilePath).wait();
+					if(data.projectVersion && data.projectVersion.toString() !== "1") {
+						this.$errors.fail("FUTURE_PROJECT_VER");
```
Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: This errors means nothing for the users. Please use something like:
```TypeScript
this.$errors.failWithoutHelp("You have modified projectVersion property in your .abproject file. This property is used internally, please set it to 1 and try again.");
```

Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: Apf, ignore me, I didn't notice the catch later in the code.

File: appbuilder/project/project-base.ts [hunk 8]
```diff
@@ -40,5 +62,92 @@ export class Project implements Project.IProjectBase {
...
+						let configMatch = path.basename(configProjectFile).match(ProjectBase.CONFIGURATION_FROM_FILE_NAME_REGEX);
+						if(configMatch && configMatch.length > 1) {
+							let configurationName = configMatch[1];
+							let configProjectContent = this.$fs.readJson(configProjectFile).wait();
+							this.configurationSpecificData[configurationName.toLowerCase()] = <any>_.merge(JSON.parse(JSON.stringify(this.projectData)), configProjectContent);
+							this._hasBuildConfigurations = true;
+						}
+					});
+				} catch(err) {
+					if(err === "FUTURE_PROJECT_VER") {
```
Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: err is never "FUTURE_PROJECT_VER", this should be:
```TypeScript
if (err.message === "FUTURE_PREOJECT_VER") {
```

File: appbuilder/project/project-base.ts [hunk 9]
```diff
@@ -40,5 +62,92 @@ export class Project implements Project.IProjectBase {
...
+						this.$errors.fail({
+							formatStr: "This project is created by a newer version of AppBuilder. Upgrade AppBuilder CLI to work with it.",
+							suppressCommandHelp: true
+						});
+					}
+					this.$errors.fail({
```
Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: this should be `failWithoutHelp`

File: appbuilder/project/project-base.ts [hunk 10]
```diff
@@ -40,5 +62,92 @@ export class Project implements Project.IProjectBase {
...
+						if(configMatch && configMatch.length > 1) {
+							let configurationName = configMatch[1];
+							let configProjectContent = this.$fs.readJson(configProjectFile).wait();
+							this.configurationSpecificData[configurationName.toLowerCase()] = <any>_.merge(JSON.parse(JSON.stringify(this.projectData)), configProjectContent);
+							this._hasBuildConfigurations = true;
+						}
+					});
+				} catch(err) {
+					if(err === "FUTURE_PROJECT_VER") {
+						this.$errors.fail({
```
Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: `failWithoutHelp`

File: appbuilder/project/project-base.ts (hunk 1 above)
Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: what is `this.projectData` and why do we have to stringify it and parse it after that ?

File: mobile/ios/device/ios-application-manager.ts [hunk 11]
```diff
@@ -93,14 +104,16 @@ export class IOSApplicationManager extends ApplicationManagerBase {
...
+							"configuration": "live"
 						},
 						{
 							"CFBundleIdentifier": "com.ionic.viewapp"
 						},
 						{
 							"IceniumLiveSyncEnabled": true,
-							"CFBundleIdentifier": "com.telerik.samplepinchandzoom"
+							"CFBundleIdentifier": "com.telerik.samplepinchandzoom",
+							configuration: "test"
```
Comment (rosen-vladimirov) [inline mobile/ios/device/ios-application-manager.ts]: missing quotes around configuration

File: services/project-files-provider-base.ts [hunk 12]
```diff
@@ -9,14 +10,15 @@ export abstract class ProjectFilesProviderBase implements IProjectFilesProvider
...
 		return path.join(path.dirname(filePath), projectFileInfo.onDeviceFileName);
 	}
 
-	public getProjectFileInfo(filePath: string, platform?: string): IProjectFileInfo {
+	public getProjectFileInfo(filePath: string, platform: string, projectFilesConfig?: IProjectFilesConfig): IProjectFileInfo {
 		let parsed = this.parseFile(filePath, this.$mobileHelper.platformNames, platform || "");
+		let basicConfigurations = [Configurations.Debug.toLowerCase(), Configurations.Release.toLowerCase()];
 		if (!parsed) {
-			parsed = this.parseFile(filePath, ["debug", "release"], this.$options.release ? "release" : "debug");
+			parsed = this.parseFile(filePath, basicConfigurations.concat(projectFilesConfig && projectFilesConfig.additionalConfigurations || []), projectFilesConfig && projectFilesConfig.configuration || basicConfigurations[0]);
```
Comment (rosen-vladimirov) [inline services/project-files-provider-base.ts]: can you extract the `basicConfigurations.concat(projectFilesConfig && projectFilesConfig.additionalConfigurations || []` and `projectFilesConfig && projectFilesConfig.configuration || basicConfigurations[0]` in separate variables in order to simplify the call to `parseFile`

File: mobile/android/android-device-file-system.ts [hunk 13]
```diff
@@ -14,8 +14,13 @@ export class AndroidDeviceFileSystem implements Mobile.IDeviceFileSystem {
 		private $injector: IInjector,
 		private $options: ICommonOptions) { }
 
-	public listFiles(devicePath: string): IFuture<void> {
-		return Future.fromResult();
+	public listFiles(devicePath: string, appIdentifier?: string): IFuture<any> {
+		let listCommandArgs = ["ls", "-al", devicePath];
```
Comment (rosen-vladimirov) [inline mobile/android/android-device-file-system.ts]: are you sure you have `-al` options on all Android devices. Have you tested this on Samsung for example?

File: appbuilder/project/project-base.ts (hunk 1 above)
Comment (Mitko-Kerezov) [inline appbuilder/project/project-base.ts]: @TsvetanMilanov 
We don't want case sensitivity with the project configurations

@rosen-vladimirov 
stringify and parse is actually a way of obtaining a deep clone of an object - I'll extract this into a method in helpers

File: appbuilder/project/project-base.ts (hunk 9 above)
Comment (Mitko-Kerezov) [inline appbuilder/project/project-base.ts]: actually it doesn't display any help because `suppressCommandHelp` is passed as `true`, however I'll still rewrite it as a call to `failWithoutHelp`

File: mobile/android/android-device-file-system.ts (hunk 13 above)
Comment (Mitko-Kerezov) [inline mobile/android/android-device-file-system.ts]: changing this to `-a` option only. This option **should** be available on all android devices

**Comment by justcodebuilduser** [comment]
:heart:

File: helpers.ts [hunk 14]
```diff
@@ -271,6 +271,10 @@ export function whenAny<T>(...futures: IFuture<T>[]): IFuture<IFuture<T>> {
 	return resultFuture;
 }
 
+export function getDeepClone(obj: any): any {
```
Comment (rosen-vladimirov) [inline helpers.ts]: remove this and use `_.deepClone` as discussed

File: appbuilder/project/project-base.ts [hunk 15]
```diff
@@ -1,29 +1,50 @@
...
+		return this._projectData;
+	}
 
-		return null;
+	public set projectData(projectData: Project.IData) {
+		this._projectData = projectData;
+	}
+
+	public projectDir: string;
+	public getProjectDir(): IFuture<string> {
```
Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: can't we remove this?

File: appbuilder/project/project-base.ts [hunk 16]
```diff
@@ -40,5 +61,82 @@ export class Project implements Project.IProjectBase {
...
+	public get hasBuildConfigurations(): boolean {
+		return this._hasBuildConfigurations;
+	}
+
+	public get projectInformation(): Project.IProjectInformation {
+		return {
+			projectData: this.projectData,
+			configurationSpecificData: this.configurationSpecificData,
+			hasBuildConfigurations: this.hasBuildConfigurations,
+			configurations: Object.keys(this.configurationSpecificData)
```
Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: `_.keys` is safer

File: appbuilder/project/project-base.ts [hunk 17]
```diff
@@ -40,5 +61,82 @@ export class Project implements Project.IProjectBase {
...
+	protected abstract saveProjectIfNeeded(): void;
+
+	protected readProjectData(): IFuture<void> {
+		return (() => {
+			let projectDir = this.getProjectDir().wait();
+			this.setShouldSaveProject(false);
+			if(projectDir) {
+				let projectFilePath = path.join(projectDir, this.$projectConstants.PROJECT_FILE);
+				try {
+					let data = this.$fs.readJson(projectFilePath).wait();
```
Comment (rosen-vladimirov) [inline appbuilder/project/project-base.ts]: this method is waaay too long, can you extract the first part to a method, for example `readProjectFile`:
```TypeScript
let projectFilePath = path.join(projectDir, this.$projectConstants.PROJECT_FILE);
try {
	let { data, shouldSaveProject } = this.readProjectFile(projectFilePath);
	this.projectData = data;
	this.setShouldSaveProject(shouldSaveProject);

	this.validate();
....
}
```

File: mobile/ios/simulator/ios-simulator-application-manager.ts [hunk 18]
```diff
@@ -55,8 +55,25 @@ export class IOSSimulatorApplicationManager extends ApplicationManagerBase {
 		return true;
 	}
 
-	public isLiveSyncSupported(appIdentifier: string): IFuture<boolean> {
-		return ((): boolean => {
+	public getApplicationInfo(applicationIdentifier: string): IFuture<Mobile.IApplicationInfo> {
+		return ((): Mobile.IApplicationInfo => {
+			let applicationPath = this.iosSim.getApplicationPath(this.identifier, applicationIdentifier),
```
Comment (rosen-vladimirov) [inline mobile/ios/simulator/ios-simulator-application-manager.ts]: this method shares a lot of code with `isLiveSyncSupported` method below. Can you extract common method that both methods will call?

File: mobile/ios/simulator/ios-simulator-application-manager.ts [hunk 19]
```diff
@@ -55,8 +55,25 @@ export class IOSSimulatorApplicationManager extends ApplicationManagerBase {
...
+					deviceIdentifier: this.identifier,
+					configuration: plistContent && plistContent.configuration
+				};
+			}
+
+			return result;
+		}).future<Mobile.IApplicationInfo>()();
+	}
+
+	public isLiveSyncSupported(appIdentifier: string): IFuture<boolean> {		return ((): boolean => {
```
Comment (rosen-vladimirov) [inline mobile/ios/simulator/ios-simulator-application-manager.ts]: hm, new line maybe?

File: helpers.ts (hunk 14 above)
Comment (Mitko-Kerezov) [inline helpers.ts]: `_.cloneDeep` maybe 😼

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: appbuilder/project/project-base.ts (hunk 15 above)
Comment (Mitko-Kerezov) [inline appbuilder/project/project-base.ts]: I think we can, but it'd have to be on a separate PR - this method is widely used throughout the code and a change like that would bring about a lot of unnecessary noise for this PR

**Comment by justcodebuilduser** [comment]
:heart:

File: options.ts [hunk 20]
```diff
@@ -46,32 +46,34 @@ export class OptionsBase {
...
+			ddi: { type: OptionType.String }, // the path to developer  disk image
+			justlaunch: { type: OptionType.Boolean },
+			file: { type: OptionType.String },
+			force: { type: OptionType.Boolean, alias: "f" },
+			companion: { type: OptionType.Boolean },
+			emulator: { type: OptionType.Boolean },
+			sdk: { type: OptionType.String },
+			template: { type: OptionType.String },
+			release: { type: OptionType.Boolean, alias: "r" },
+			var: {type: OptionType.Object },
```
Comment (rosen-vladimirov) [inline options.ts]: space before type

**Comment by rosen-vladimirov** [comment]
:+1:

**Comment by justcodebuilduser** [comment]
:heart:
